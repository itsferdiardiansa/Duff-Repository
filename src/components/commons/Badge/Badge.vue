<template>
  <div
    :class="customClass"
    :title="title"
  >
    <div :class="`${prefixClass}-badge--wrapper`" v-if="!dot">
      <template v-if="icon.length">
        <div :class="`${prefixClass}-badge--icon`">
          <font-awesome-icon :icon="icon" />
        </div>
      </template>

      <template v-if="hasSlot || label.length">
        <div :class="[`${prefixClass}-badge--label`, {'has-icon': icon.length}]">
          <slot>
            {{ label }}
          </slot>
        </div>
      </template>
    </div>
  </div>
</template>
<script>
import { computed, getCurrentInstance } from 'vue'

export default {
  props: {
    label: {
      type: String,
      default: '',
    },
    title: {
      type: String,
      default: '',
    },
    tooltip: {
      type: String,
      default: ''
    },
    inverse: {
      type: Boolean,
      default: false
    },
    variant: {
      type: String,
      default: 'light',
      validator: function(variant) {
        return ~['primary', 'danger', 'warning', 'dark', 'success', 'light'].indexOf(variant)
      }
    },
    size: {
      type: String,
      default: 'base',
      validator: function(variant) {
        return ~['xs', 'sm', 'base', 'lg', 'xl'].indexOf(variant)
      }
    },
    icon: {
      type: Array,
      default: () => []
    },
    bold: {
      type: Boolean,
      default: true,
    },
    rounded: {
      type: Boolean,
      default: true,
    },
    pill: {
      type: Boolean,
      default: false
    },
    dot: {
      type: Boolean,
      default: false
    }
  },
  setup(props, { slots }) {
    let root = getCurrentInstance()

    const getVariantClass = (prefixBtnClass) => {
      const { inverse, variant } = props 
      let btnClass = `${prefixBtnClass}--${variant}`

      if(inverse) btnClass += '-inverse'
      
      return btnClass
    }

    const customClass = computed(() => {
      const { data: { prefixClass } } = root
      const { size, rounded, bold, pill, dot, label } = props
      const btnClass = prefixClass.concat('-badge')
      let _class = [btnClass]

      _class.push(getVariantClass(btnClass))

      if(size) _class.push(btnClass.concat('--' + size))

      if (bold) _class.push('text-bold')

      if(pill)
        _class.push('pill')
      else
        if (!rounded) _class.push('no-rounded')

      if(dot && (!hasSlot.value && !label.length)) _class.push('dot')

      return [_class.join(' ')]
    })

    const hasSlot = computed(() => (
      !!slots.default
    ))

    return {
      customClass,
      hasSlot
    }
  },
}
</script>
<style lang="scss">
@import './variant.scss';
@import './size.scss';

.#{$prefixClass}-badge {
  @apply h-9 focus:outline-none rounded px-3 transition duration-300 relative inline-block;
  @apply text-gray-800 overflow-hidden whitespace-nowrap;

  @include variant;

  @include size;

  &.no-rounded {
    @apply rounded-none;
  }

  &.pill {
    @apply rounded-full;
  }

  &.text-bold {
    @apply font-bold;
  }

  &.disabled {
    @apply disabled:opacity-70 pointer-events-none;
  }
  
  &--wrapper {
    @apply h-full flex-grow-0 flex flex-wrap items-center justify-center;
  }

  &--label {
    @apply pt-0.5;

    &.has-icon {
      @apply ml-2;
    }
  }
}
</style>