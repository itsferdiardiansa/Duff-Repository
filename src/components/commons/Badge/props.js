export default {
  label: {
    type: [String, Number],
    default: '',
  },
  inverse: {
    type: Boolean,
    default: false,
  },
  variant: {
    type: String,
    default: 'light',
    validator: function (variant) {
      return ~[
        'primary',
        'danger',
        'warning',
        'dark',
        'success',
        'light',
        'orange',
      ].indexOf(variant);
    },
  },
  size: {
    type: String,
    default: 'xs',
    validator: function (variant) {
      return ~['xs', 'sm', 'base', 'lg', 'xl'].indexOf(variant);
    },
  },
  icon: {
    type: Array,
    default: () => [],
  },
  bold: {
    type: Boolean,
    default: true,
  },
  rounded: {
    type: Boolean,
    default: true,
  },
  pill: {
    type: Boolean,
    default: false,
  },
  dot: {
    type: Boolean,
    default: false,
  },
};
