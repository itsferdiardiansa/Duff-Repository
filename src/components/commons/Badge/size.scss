@mixin size {
  &--xs {
    @apply h-6 text-xs px-2;
    
    &.dot {
      @apply h-4;
    }

    svg {
      @apply h-2.5;
    }
  }

  &--sm {
    @apply h-7 text-sm px-2.5;

    &.dot {
      @apply h-5;
    }

    svg {
      @apply h-3;
    }
  }

  &--base {
    @apply h-9 px-3;

    &.dot {
      @apply h-6;
    }

    svg {
      @apply h-3.5;
    }
  }

  &--lg {
    @apply h-12 text-lg px-4;

    &.dot {
      @apply h-8;
    }

    svg {
      @apply h-4;
    }
  }

  &--xl {
    @apply h-14 text-xl px-5;

    &.dot {
      @apply h-10;
    }

    svg {
      @apply h-5 mr-2;
    }
  }
}