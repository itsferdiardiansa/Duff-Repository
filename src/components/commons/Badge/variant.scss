@mixin variant {
  &.primary {
    @apply bg-primary text-gray-200;
  }

  &.danger {
    @apply bg-danger text-gray-200;
  }

  &.warning {
    @apply bg-warning text-gray-800;
  }

  &.success {
    @apply bg-success text-gray-200;
  }

  &.dark {
    @apply bg-dark text-gray-200;
  }
}
