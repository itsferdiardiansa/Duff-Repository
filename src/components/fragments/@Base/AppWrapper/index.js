/* eslint-disable no-unused-vars */
import {
  defineComponent,
  h,
  ref,
  getCurrentInstance,
  unref,
  computed,
} from 'vue'
import { useStore } from 'vuex'
import { useRouter } from 'vue-router'
import globalSetting from '@plugin/globalSetting'
import ErrorComponent from '@base/Error'
import subscribeFormAction from './subscribeFormAction'
import asyncModuleLoad from './asyncModuleLoad'

const AppWrapper = defineComponent({
  beforeCreate() {
    const {
      appContext: { app },
    } = getCurrentInstance()
    const store = useStore()
    const router = useRouter()

    store.subscribe(({ payload }, state) => {
      subscribeFormAction(payload, router)
    })

    // Handle async module load
    asyncModuleLoad()

    // register global mixin
    app.mixin(globalSetting)
  },
  beforeRouteEnter() {
    console.log('before router enter wrapper')
  },
  errorCaptured(error, component) {},
  setup(props, { slots }) {
    const store = useStore()

    const errorChunkModule = computed(() => store.getters['app/getChunks'])

    return () => {
      return unref(errorChunkModule).onError
        ? h(ErrorComponent, {})
        : slots.default()
    }
  },
})

export default AppWrapper
