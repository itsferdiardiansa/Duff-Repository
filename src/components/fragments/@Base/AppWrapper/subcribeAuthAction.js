export default (type, router) => {
  // Router push
  if (type === 'user/setToken') router.push('/');
};
