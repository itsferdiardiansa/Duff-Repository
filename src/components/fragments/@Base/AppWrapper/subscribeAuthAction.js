const alert = (content, variant) => sAlert.show({ content, variant });

export default (type, router) => {
  switch (type) {
    case 'user/loginSuccess':
      alert('Welcome aboard', 'success');
      break;
    case 'user/loginFailed':
      alert('Failed to login', 'danger');
      break;
    case 'user/removeAuth':
      alert('Thank you!', 'success');

      router.push('/login');
      break;
    default:
      break;
  }
};
