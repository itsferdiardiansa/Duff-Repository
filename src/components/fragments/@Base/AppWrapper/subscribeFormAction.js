export default (type, payload, router) => {
  if (
    ['form.create', 'form.update', 'form.delete'].includes(
      payload?.requestData?.action
    )
  ) {
    const variant = type.match(/fetchFailed/g) ? 'danger' : 'success';
    const content = payload?.responseData?.message;

    sAlert.show({ variant, content });

    if (
      ['form.create', 'form.update'].includes(payload?.requestData?.action) &&
      variant !== 'danger' &&
      Reflect.has(payload?.requestData, 'redirectUrl')
    ) {
      router.push(payload?.requestData.redirectUrl);
    }
  }
};
