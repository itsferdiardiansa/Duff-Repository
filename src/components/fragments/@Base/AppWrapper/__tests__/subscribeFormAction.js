import { mount } from '@vue/test-utils';
import { createStore } from 'vuex';
import AppWrapper from '@base/AppWrapper';
import Badge from '@common/Badge';
import heroModule from '@store/modules/hero';

describe('fragment/AppWrapper/subscribeFormAction', () => {
  let Component, store, alertElement, prefixClass;

  beforeAll(() => {
    store = createStore({
      modules: {
        hero: heroModule,
      },
    });

    Component = mount(
      <AppWrapper>
        <Badge label="its_ferdi" variant="success" />
      </AppWrapper>,
      {
        global: {
          plugins: [store],
        },
      }
    );

    prefixClass = Component.rootVM.prefixClass;
  });

  it('will render children and match snapshot', () => {
    expect(Component.html()).toMatchSnapshot();
  });

  it('will not display a success alert when the request body has not action properties', async () => {
    const data = { title: 'Its ferdi' };

    await Component.vm.$store.dispatch('hero/postData', {
      redirectUrl: '/hero',
      data,
    });

    expect(
      document.body.querySelector(`.${Component.rootVM.prefixClass}-alert`)
    ).toBeFalsy();
    expect(document.body).toMatchSnapshot();
  });

  it('will display a success alert when the request is successful', async () => {
    const data = { title: 'Its ferdi' };

    await Component.vm.$store.dispatch('hero/postData', {
      action: 'form.create',
      redirectUrl: '/hero',
      data,
    });

    expect(document.body).toMatchSnapshot();
  });

  it('will display a danger alert when the request fails', async () => {
    const Component = mount(
      <AppWrapper>
        <ActionForm />
      </AppWrapper>,
      {
        global: {
          plugins: [store],
        },
      }
    );

    mockCreateHero.mockReset();
    mockCreateHero.mockRejectedValue({
      message: 'internal.error',
      result: {},
    });

    await Component.vm.$store.dispatch('hero/postData', {
      action: 'form.create',
    });

    expect(
      document.body.querySelector(
        `.${Component.rootVM.prefixClass}-alert--danger`
      )
    ).toBeTruthy();
    expect(document.body).toMatchSnapshot();
  });
});
