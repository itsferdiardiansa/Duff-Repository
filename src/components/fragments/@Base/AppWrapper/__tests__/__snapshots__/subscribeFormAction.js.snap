// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fragment/AppWrapper/subscribeFormAction will display a danger alert when the request fails 1`] = `
<body>
  
  
  <div
    id="app"
  >
    
  
    <div
      <=""
      data-v-app=""
      div=""
      id="plugin__12e5"
    >
      
      <div
        class="ss-alert ss-alert--success slide-top-enter-from slide-top-enter-active"
      >
        <div
          class="ss-alert--icon"
        >
          <svg
            class=""
            viewBox="0 0 512 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M504 256c0 136.967-111.033 248-248 248S8 392.967 8 256 119.033 8 256 8s248 111.033 248 248zM227.314 387.314l184-184c6.248-6.248 6.248-16.379 0-22.627l-22.627-22.627c-6.248-6.249-16.379-6.249-22.628 0L216 308.118l-70.059-70.059c-6.248-6.248-16.379-6.248-22.628 0l-22.627 22.627c-6.248 6.248-6.248 16.379 0 22.627l104 104c6.249 6.249 16.379 6.249 22.628.001z"
              fill="currentColor"
            />
          </svg>
        </div>
        <div
          class="ss-alert--wrapper"
        >
          <!--v-if-->
          <!--v-if-->
          <div
            class="ss-alert--content"
          >
            <label>
              cms.hero.create.success
            </label>
          </div>
        </div>
      </div>
      <div
        class="ss-alert ss-alert--danger slide-top-enter-from slide-top-enter-active"
      >
        <div
          class="ss-alert--icon"
        >
          <svg
            class=""
            viewBox="0 0 512 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M256 8C119 8 8 119 8 256s111 248 248 248 248-111 248-248S393 8 256 8zm121.6 313.1c4.7 4.7 4.7 12.3 0 17L338 377.6c-4.7 4.7-12.3 4.7-17 0L256 312l-65.1 65.6c-4.7 4.7-12.3 4.7-17 0L134.4 338c-4.7-4.7-4.7-12.3 0-17l65.6-65-65.6-65.1c-4.7-4.7-4.7-12.3 0-17l39.6-39.6c4.7-4.7 12.3-4.7 17 0l65 65.7 65.1-65.6c4.7-4.7 12.3-4.7 17 0l39.6 39.6c4.7 4.7 4.7 12.3 0 17L312 256l65.6 65.1z"
              fill="currentColor"
            />
          </svg>
        </div>
        <div
          class="ss-alert--wrapper"
        >
          <!--v-if-->
          <!--v-if-->
          <div
            class="ss-alert--content"
          >
            <label>
              internal.error
            </label>
          </div>
        </div>
      </div>
      <div
        class="ss-alert ss-alert--danger slide-top-enter-from slide-top-enter-active"
      >
        <div
          class="ss-alert--icon"
        >
          <svg
            class=""
            viewBox="0 0 512 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M256 8C119 8 8 119 8 256s111 248 248 248 248-111 248-248S393 8 256 8zm121.6 313.1c4.7 4.7 4.7 12.3 0 17L338 377.6c-4.7 4.7-12.3 4.7-17 0L256 312l-65.1 65.6c-4.7 4.7-12.3 4.7-17 0L134.4 338c-4.7-4.7-4.7-12.3 0-17l65.6-65-65.6-65.1c-4.7-4.7-4.7-12.3 0-17l39.6-39.6c4.7-4.7 12.3-4.7 17 0l65 65.7 65.1-65.6c4.7-4.7 12.3-4.7 17 0l39.6 39.6c4.7 4.7 4.7 12.3 0 17L312 256l65.6 65.1z"
              fill="currentColor"
            />
          </svg>
        </div>
        <div
          class="ss-alert--wrapper"
        >
          <!--v-if-->
          <!--v-if-->
          <div
            class="ss-alert--content"
          >
            <label>
              internal.error
            </label>
          </div>
        </div>
      </div>
      
    </div>
  </div>
  <div
    class="ss-alert-container"
    id="plugin__12e5"
    role="alert"
  />
</body>
`;

exports[`fragment/AppWrapper/subscribeFormAction will display a success alert when the request is successful 1`] = `
<body>
  
  
  <div
    id="app"
  >
    
  
    <div
      <=""
      data-v-app=""
      div=""
      id="plugin__12e5"
    >
      
      <div
        class="ss-alert ss-alert--success slide-top-enter-from slide-top-enter-active"
      >
        <div
          class="ss-alert--icon"
        >
          <svg
            class=""
            viewBox="0 0 512 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M504 256c0 136.967-111.033 248-248 248S8 392.967 8 256 119.033 8 256 8s248 111.033 248 248zM227.314 387.314l184-184c6.248-6.248 6.248-16.379 0-22.627l-22.627-22.627c-6.248-6.249-16.379-6.249-22.628 0L216 308.118l-70.059-70.059c-6.248-6.248-16.379-6.248-22.628 0l-22.627 22.627c-6.248 6.248-6.248 16.379 0 22.627l104 104c6.249 6.249 16.379 6.249 22.628.001z"
              fill="currentColor"
            />
          </svg>
        </div>
        <div
          class="ss-alert--wrapper"
        >
          <!--v-if-->
          <!--v-if-->
          <div
            class="ss-alert--content"
          >
            <label>
              cms.hero.create.success
            </label>
          </div>
        </div>
      </div>
      
    </div>
  </div>
  <div
    class="ss-alert-container"
    id="plugin__12e5"
    role="alert"
  />
</body>
`;

exports[`fragment/AppWrapper/subscribeFormAction will not display a success alert when the request body has not action properties 1`] = `
<body>
  
  
  <div
    id="app"
  >
    
  
    <div
      <=""
      data-v-app=""
      div=""
      id="plugin__12e5"
    >
      
      
    </div>
  </div>
  <div
    class="ss-alert-container"
    id="plugin__12e5"
    role="alert"
  />
</body>
`;

exports[`fragment/AppWrapper/subscribeFormAction will render children and match snapshot 1`] = `<span class="ss-badge ss-badge--success ss-badge--xs text-bold"><div class="ss-badge--wrapper"><!--v-if--><div class="ss-badge--label">its_ferdi</div></div></span>`;
